### PR TITLE
[docs] LLM inference

### PR DESCRIPTION
**Pull request by stevhliu**
This PR creates a lightweight LLM inference optimization guide focused on just providing a brief explanation and copy/paste-able code examples so users can instantly get going without necessarily needing to get too in-depth in the underlying conceptual details about how it works. It includes:

- static kv-cache + torch.compile
- speculative and prompt lookup decoding
- attention optimizations (FlashAttention-2 and SDPA)
- quantization

### CODE REVIEW

File: docs/source/en/llm_optims.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+input_text = "Planck's constant is"
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+
+outputs = compiled_model.generate(**input_ids)
+```
+
+</hfoption>
+<hfoption id="setup_cache">
+
+Access the model's `_set_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.
```
Comment (tengomucho) [inline docs/source/en/llm_optims.md:70]: I think it would be great to think about changing the method name from `_set_cache` to `set_cache` if it is not supposed to be internal.
Also, I think it would be good to explain what is the value that should be put in the `max_cache_len` (maximum number of tokens if I got it right).

File: docs/source/en/llm_optims.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    "google/gemma-2b",
+    torch_dtype=bfloat16,
+    attn_implementation="flash_attention_2",
+    load_in_8bit=True,
+)
+```
+
+### PyTorch scaled dot product attention
+
+Scaled dot product attention (SDPA) is automatically enabled in PyTorch 2.0 and it supports FlashAttention, xFormers, and a C++ PyTorch implementation. SDPA chooses the most performant attention algorithm based on your hardware, but you could also explicitly enable or disable the attention algorithm you want to use.
```
Comment (tengomucho) [inline docs/source/en/llm_optims.md:213]: You say "SDPA chooses the most performan algorithm based on your hardware", but that is true for CUDA, not other backends (mps, xla, etc), right?

Comment (stevhliu) [reply docs/source/en/llm_optims.md:213]: Correct, I'll update and clarify this. Thanks :)

File: docs/source/en/llm_optims.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+</hfoption>
+</hfoptions>
+
+## Attention optimizations
+
+A known issue with transformer models is that the self-attention mechanism grows quadratically in compute and memory with the number of input tokens. This limitation is only magnified in LLMs which can handle much longer sequences. To address this, try FlashAttention2 or PyTorch's scaled dot product attention (SDPA), which are more memory efficient implementations and can accelerate inference.
+
+### FlashAttention-2
+
+FlashAttention and [FlashAttention-2](./perf_infer_gpu_one#flashattention-2) breaks up the attention computation into smaller chunks and reduces the number of intermediate read/write operations to GPU memory to speed up inference. FlashAttention-2 improves on the original FlashAttention algorithm by also parallelizing over sequence length dimension and better partitioning work on the hardware to reduce synchronization and communication overhead.
```
Comment (khipp) [inline docs/source/en/llm_optims.md:196]:
```suggestion
FlashAttention and [FlashAttention-2](./perf_infer_gpu_one#flashattention-2) break up the attention computation into smaller chunks and reduce the number of intermediate read/write operations to GPU memory to speed up inference. FlashAttention-2 improves on the original FlashAttention algorithm by also parallelizing over sequence length dimension and better partitioning work on the hardware to reduce synchronization and communication overhead.
```

**Comment by stevhliu** [comment]
For each of the topics covered, I was wondering if:

- @ArthurZucker, would you mind having a look at the [static kv + torch.compile](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29791/en/llm_optims#static-kv-cache-and-torchcompile) section?
- @gante, would you mind having a look at the [speculative decoding](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29791/en/llm_optims#speculative-decoding) section?
- @younesbelkada or @SunMarc, would you mind having a look at the [attention optimization](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29791/en/llm_optims#attention-optimizations) and [quantization](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29791/en/llm_optims#quantization) sections?

File: docs/source/en/llm_optims.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+import torch
+
+tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")
+model = AutoModelForCausalLM.from_pretrained(
+    "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto",
+).to("cuda")
+```
+
+To load a quantized model (8-bit or 4-bit) for inference, try [bitsandbytes](https://hf.co/docs/bitsandbytes) and set the `load_in_4bit` or `load_in_8bit` parameters to `True`. Loading the model in 8-bits only costs 6.87 GB of memory.
+
```
Comment (SunMarc) [inline docs/source/en/llm_optims.md:262]: Since we are deprecating `load_in_4bit` and `load_in_8bit`, I suggest using `quantization_config` instead and pass a `BitsAndBytesConfig` object.
If the user wants to do 8-bit quantization, he needs to pass `BitsAndBytesConfig(load_in_8bit=True)`. For 4-bit quantization, he needs to pass `BitsAndBytesConfig(load_in_4bit=True)`.

File: docs/source/en/llm_optims.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+```py
+from transformers import AutoTokenizer, AutoModelForCausalLM
+import torch
+
+tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")
+model = AutoModelForCausalLM.from_pretrained(
+    "mistralai/Mistral-7B-v0.1", device_map="auto", load_in_8bit=True,
+).to("cuda")
```
Comment (SunMarc) [inline docs/source/en/llm_optims.md:270]: We shouldn't move the model when we use device_map="auto" since the model is already dispatch the GPUs. It might trigger some issues in the multi-gpu case otherwise. 
```suggestion
model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", device_map="auto", load_in_8bit=True,
)
```

File: docs/source/en/llm_optims.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+To load Mistral-7B-v0.1 in half-precision, set the `torch_dtype` parameter in the [`~transformers.AutoModelForCausalLM.from_pretrained`] method to `torch.bfloat16`. This only costs 13.74GB of memory.
+
+```py
+from transformers import AutoTokenizer, AutoModelForCausalLM
+import torch
+
+tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")
+model = AutoModelForCausalLM.from_pretrained(
+    "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto",
+).to("cuda")
```
Comment (SunMarc) [inline docs/source/en/llm_optims.md:258]: same issue
```suggestion
model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto",
)
```

**Review by SunMarc — APPROVED** [review]
Thanks for the great work @stevhliu ! I left a couple of suggestions.

File: docs/source/en/llm_optims.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+```
+
+Call torch.compile on the model to compile it with the static kv-cache.
+
+```py
+compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
+input_text = "Planck's constant is"
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+
+outputs = compiled_model.generate(**input_ids)
```
Comment (ArthurZucker) [inline docs/source/en/llm_optims.md:64]: generate is not yet compatible with torch compile so would only recommend the forward pass to be shown here! And talk about 4x!

File: docs/source/en/llm_optims.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+## Static kv-cache and torch.compile
+
+During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time since the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.
+
+To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses Python code into fast and optimized kernels.
+
+The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
+
+> [!WARNING]
+> Currently, only [Gemma](./model_doc/gemma) and [Llama](./model_doc/llama2) models support static kv-cache and torch.compile.
```
Comment (ArthurZucker) [inline docs/source/en/llm_optims.md:32]: and cohere I think

**Review by ArthurZucker — COMMENTED** [review]
Thanks!

File: docs/source/en/llm_optims.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
+input_text = "Planck's constant is"
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+outputs = compiled_model(input_ids)
+```
+
+</hfoption>
+</hfoptions>
+
+## Speculative decoding
```
Comment (gante) [inline docs/source/en/llm_optims.md:88]: Shameless plug: https://huggingface.co/blog/assisted-generation 😉

**Review by gante — APPROVED** [review]
Very cool 🙌

**Review by younesbelkada — APPROVED** [review]
Fantastic work ! 🚀

File: docs/source/en/llm_optims.md [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+
+def decode_one_tokens(model, cur_token, input_pos, cache_position):
+    logits = model(
+        cur_token, position_ids=input_pos, cache_position=cache_position, return_dict=False, use_cache=True
+    )[0]
+    new_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+    return new_token
+```
+
+The are two important things you must do to enable static kv-cache and torch.compile with the `_setup_cache` method:
```
Comment (amyeroberts) [inline docs/source/en/llm_optims.md:101]: I'm not sure about advertising using private methods - this means they're effectively public and need to have the same strict backwards compatibility handling of public methods

File: docs/source/en/llm_optims.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+     generated_ids[:, cache_position] = inputs["input_ids"].to(torch_device).to(torch.int)
+
+     logits = model(**inputs, cache_position=cache_position, return_dict=False, use_cache=True)[0]
+     next_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+     generated_ids[:, seq_length] = next_token[:, 0]
+
+     decode_one_tokens = torch.compile(decode_one_tokens, mode="reduce-overhead", fullgraph=True)
+     cache_position = torch.tensor([seq_length + 1], device=torch_device)
+     for _ in range(1, NUM_TOKENS_TO_GENERATE):
+         with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=False, enable_math=True):
```
Comment (amyeroberts) [inline docs/source/en/llm_optims.md:124]: It would be good to have a comment here explaining the value of this context manager

File: docs/source/en/llm_optims.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+             next_token = decode_one_tokens(model, next_token.clone(), None, cache_position)
+             generated_ids[:, cache_position] = next_token.int()
+         cache_position += 1
+
+text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
+text
+['Simply put, the theory of relativity states that 1) the speed of light is constant, 2) the speed of light is the same for all observers, and 3) the laws of physics are the same for all observers.',
+ 'My favorite all time favorite condiment is ketchup. I love it on everything. I love it on my eggs, my fries, my chicken, my burgers, my hot dogs, my sandwiches, my salads, my p']
+```
```
Comment (amyeroberts) [inline docs/source/en/llm_optims.md:133]: For both of the compile options - I think it would be really compelling for a user if they had actual numbers e.g. fdo we get a 4x speed up under all cases? Do we get the same speed-up for these two cache approaches?

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding!

File: docs/source/en/llm_optims.md (hunk 12 above)
Comment (stevhliu) [reply docs/source/en/llm_optims.md:135]: Totally agree, I'll ask @gante about this 🙂 

In the meantime, would it be ok to merge this PR first and add this in a separate PR or would you prefer to wait?

File: docs/source/en/llm_optims.md (hunk 10 above)
Comment (stevhliu) [reply docs/source/en/llm_optims.md]: I think it's ok with a warning that this is still experimental and its design will probably change in the future, so the user should be aware that some things may not work as expected, but in the meantime, you can still get excited/try this out? 

Would it help if I added a sentence about how it may not be backward compatible?

Comment (amyeroberts) [reply docs/source/en/llm_optims.md]: That works for me!

File: docs/source/en/llm_optims.md (hunk 12 above)
Comment (amyeroberts) [reply docs/source/en/llm_optims.md:135]: Sure!

File: docs/source/en/llm_optims.md [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+This guide will show you how to use the optimization techniques available in Transformers to accelerate LLM inference.
+
+> [!TIP]
+> Hugging Face also provides [Text Generation Inference (TGI)](https://hf.co/docs/text-generation-inference), a library dedicated to deploying and serving highly optimized LLMs for inference. It includes more optimization features not included in Transformers, such as continuous batching for increasing throughput and tensor parallelism for multi-GPU inference.
+
+## Static kv-cache and torch.compile
+
+During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time because the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.
+
+To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses PyTorch code into fast and optimized kernels.
```
Comment (ArthurZucker) [inline docs/source/en/llm_optims.md:27]:
```suggestion
To optimize this, you can use a kv-cache to store the past keys and values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses PyTorch code into fast and optimized kernels.
```

File: docs/source/en/llm_optims.md [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+
+The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
```
Comment (ArthurZucker) [inline docs/source/en/llm_optims.md:29]:
```suggestion
The *static kv-cache* solves this issue by pre-allocating the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
```

File: docs/source/en/llm_optims.md (hunk 12 above)
Comment (ArthurZucker) [reply docs/source/en/llm_optims.md:135]: Sure

**Review by ArthurZucker — APPROVED** [review]
Sorry for the late review here!

**Review by tengomucho — APPROVED** [review]
